### ISSUE TITLE
0.7.0 regression

### ISSUE THREAD

**Issue by ncave** [fable-compiler/Fable#503]
This is compiling on master branch but not on 0.7.0 (this is a cut-down minimal reproduction, so don't read too much into the code):
```
    open System.Collections.Generic

    type MapTree<'Key,'T> = 
        | MapEmpty 
        | MapNode of 'Key * 'T * MapTree<'Key,'T> *  MapTree<'Key,'T> * int

    type MapIterator<'Key,'T>(s:MapTree<'Key,'T>) = 
        let rec collapseLHS stack =
            match stack with
            | []                           -> []
            | MapEmpty             :: rest -> collapseLHS rest
            | (MapNode(_,_,MapEmpty,MapEmpty,_)) :: _ -> stack
            | (MapNode(k,v,l,r,_)) :: rest ->
                collapseLHS (l :: MapNode(k,v,MapEmpty,MapEmpty,1) :: r :: rest)
        
        let mutable stack = collapseLHS [s]

        member i.Current =
            match stack with
                | (MapNode(k,v,MapEmpty,MapEmpty,_)) :: _ -> new KeyValuePair<_,_>(k,v)
                | _ -> failwith "error"
```
Here is the error: `ERROR: Cannot access type: System.Collections.Generic.KeyValuePair`

**Comment by ncave** [comment]
If it helps, here is the Fable AST for the above for master and 0.7.0 branches.
They are quite different, I guess the changes in 0.7.0 are not trivial:
[ast.zip](https://github.com/fable-compiler/Fable/files/551692/ast.zip)

**Comment by alfonsogarciacaro** [comment]
Interesting, it's the first time I've seen a big Fable AST printed like this 😄 

Thanks a lot for the detailed info, I'll investigate :+1:

**Comment by ncave** [comment]
That's just what sprintf "%A" produces on Fable declarations, so may not be very readable. I have added the FSharp AST as well, hopefully that helps. It's amazing how much more compact the FSharp AST is vs. the Fable AST / Babel AST, I guess that's because it's terse and more expressive.

**Comment by alfonsogarciacaro** [comment]
Sorry, I should have noticed the problem with your bug description. In 0.7.0 types include reflection info about their properties so types that were previously erased (like `KeyValuePair`) must be now referenced. I need to find a way to implement a catch-all for these cases, as there've been several reports like this with other BCL types.

At least this specific case should be fixed in the next 0.7.0 alpha to be released in a few minutes :+1:
